torch/csrc/cudnn/Descriptors.h

```diff
@@ -4,27 +4,67 @@
 #include "Exceptions.h"
 
 #include <cudnn.h>
+#include <ATen/ATen.h>
 
 namespace torch { namespace cudnn {
 
+inline int dataSize(cudnnDataType_t dataType)
+{
+  switch (dataType) {
+    case CUDNN_DATA_HALF: return 2;
+    case CUDNN_DATA_FLOAT: return 4;
+    default: return 8;
+  }
+}
+
+inline cudnnDataType_t getDataType(const at::Tensor& t) {
+  auto scalar_type = t.type().scalarType();
+  if (scalar_type == at::kFloat) {
+    return CUDNN_DATA_FLOAT;
+  } else if (scalar_type == at::kHalf) {
+    return CUDNN_DATA_HALF;
+  } else if (scalar_type == at::kDouble) {
+    return CUDNN_DATA_DOUBLE;
+  }
+  throw std::runtime_error("TensorDescriptor only supports double, float and half tensors");
+}
+
 struct TensorDescriptor
 {
   cudnnTensorDescriptor_t desc;
+
   TensorDescriptor() : desc(NULL) {
     CHECK(cudnnCreateTensorDescriptor(&desc));
   }
-  TensorDescriptor(const TensorDescriptor&) = delete;
-  TensorDescriptor(TensorDescriptor&& ref)
-  {
-    desc = ref.desc;
-    ref.desc = NULL;
+  /* implicit */ TensorDescriptor(const at::Tensor& t, int pad = 0) : desc(NULL) {
+    CHECK(cudnnCreateTensorDescriptor(&desc));
+    set(t, pad);
   }
+  TensorDescriptor(const TensorDescriptor&) = delete;
+  TensorDescriptor(TensorDescriptor&& ref) { desc = ref.desc; ref.desc = NULL; }
   ~TensorDescriptor() {
     cudnnDestroyTensorDescriptor(desc);
   }
   void set(cudnnDataType_t dataType, int dim, int* size, int* stride) {
     CHECK(cudnnSetTensorNdDescriptor(desc, dataType, dim, size, stride));
   }
+  void set(const at::Tensor &t, int pad = 0) {
+    int dim = t.ndimension();
+    if (dim > 10 || pad > 10)
```

Comment:
CUDNN_DIM_MAX is 8, defined in cudnn.h. Better to use it rather than string literal.